controller/controlplane/controller.go

```diff
@@ -714,3 +743,70 @@ func (r *Reconciler) validateControlPlaneOptions(cp *ControlPlane) (string, bool
 
 	return "", true
 }
+
+func (r *Reconciler) enforceDataPlaneNameInStatus(
+	ctx context.Context,
+	cp *gwtypes.ControlPlane,
+) (string, op.Result, error) {
+	var dataplaneName string
+	switch cp.Spec.DataPlane.Type {
+	case gwtypes.ControlPlaneDataPlaneTargetRefType:
+		dataplaneName = cp.Spec.DataPlane.Ref.Name
+	case gwtypes.ControlPlaneDataPlaneTargetManagedByType:
+		owner, hasOwner := lo.Find(cp.GetOwnerReferences(), func(owner metav1.OwnerReference) bool {
+			return strings.HasPrefix(owner.APIVersion, gatewayv1.GroupName) &&
+				owner.Kind == "Gateway"
+		})
+
+		if !hasOwner {
+			return "", op.Noop, fmt.Errorf(
+				"no Gateway owner but spec.dataplane.type set to %s",
+				gwtypes.ControlPlaneDataPlaneTargetManagedByType,
+			)
+		}
+
+		var dataplaneList operatorv1beta1.DataPlaneList
+		err := r.List(ctx, &dataplaneList, client.MatchingFields{
+			index.DataPlaneOnOwnerGatewayIndex: cp.Namespace + "/" + owner.Name,
+		})
+		if err != nil {
+			return "", op.Noop, fmt.Errorf(
+				"failed to find DataPlanes owned by Gateway %s/%s: %w",
+				cp.GetNamespace(), owner.Name, err,
+			)
+		}
+		if len(dataplaneList.Items) != 1 {
+			return "", op.Noop, fmt.Errorf(
+				"no DataPlanes owned by Gateway %s/%s",
```

Comment:
Is it possible to get more than 1 `DataPlane`s owned by `Gateway`? If so the error message may not be accurate.

Reply:
👍 97c33b8c224f52fb0fda1cecd9799d67e851abd1